[WIP] vim-patch:7.4.[1685,2163,2217,2269],8.0.00[33,40]

brcolow opened the pull request:
7.4.1685:

Problem:    There is no easy way to get all the information about a match.
Solution:   Add matchstrpos(). (Ozaki Kiichi)

https://github.com/vim/vim/commit/7fed5c18f8577b75404b80d8b9a9907b1bbd27e4

I don't think the lines here: https://github.com/vim/vim/commit/7fed5c18f8577b75404b80d8b9a9907b1bbd27e4#diff-de9ab8906fef668ed8042046e0a1eb02R15323 are needed based on the neighboring code and its difference from upstream but please advise :)

Update: Okay they are necessary but I am not sure where to put them because `list_append_string` does not return `FAIL` if the allocation was unsuccessful.

```
==32095==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 3 byte(s) in 3 object(s) allocated from:
    #0 0x4efd18 in malloc (/home/travis/build/neovim/neovim/build/bin/nvim+0x4efd18)
    #1 0xeb1534 in try_malloc /home/travis/build/neovim/neovim/src/nvim/memory.c:59:15
    #2 0xeb16f4 in xmalloc /home/travis/build/neovim/neovim/src/nvim/memory.c:93:15
    #3 0xeb1b33 in xmallocz /home/travis/build/neovim/neovim/src/nvim/memory.c:168:15
    #4 0xeb1bf8 in xmemdupz /home/travis/build/neovim/neovim/src/nvim/memory.c:186:17
    #5 0x7d0701 in list_append_string /home/travis/build/neovim/neovim/src/nvim/eval.c:5383:40
    #6 0x8f473d in find_some_match /home/travis/build/neovim/neovim/src/nvim/eval.c:12208:7
    #7 0x88e397 in f_matchstrpos /home/travis/build/neovim/neovim/src/nvim/eval.c:12530:3
    #8 0x7aa851 in call_func /home/travis/build/neovim/neovim/src/nvim/eval.c:7004:11
    #9 0x7be874 in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:6855:11
    #10 0x85043c in eval7 /home/travis/build/neovim/neovim/src/nvim/eval.c:4301:15
    #11 0x84bea5 in eval6 /home/travis/build/neovim/neovim/src/nvim/eval.c:4022:7
    #12 0x848b39 in eval5 /home/travis/build/neovim/neovim/src/nvim/eval.c:3874:7
    #13 0x8440b1 in eval4 /home/travis/build/neovim/neovim/src/nvim/eval.c:3626:7
    #14 0x843653 in eval3 /home/travis/build/neovim/neovim/src/nvim/eval.c:3548:7
    #15 0x842bf3 in eval2 /home/travis/build/neovim/neovim/src/nvim/eval.c:3485:7
    #16 0x7a3212 in eval1 /home/travis/build/neovim/neovim/src/nvim/eval.c:3418:7
    #17 0x7be59f in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:6841:9
    #18 0x7b8cc6 in ex_call /home/travis/build/neovim/neovim/src/nvim/eval.c:2804:9
    #19 0xac0841 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2197:5
    #20 0xa9f407 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #21 0x7e5300 in call_user_func /home/travis/build/neovim/neovim/src/nvim/eval.c:20896:3
    #22 0x7aa28a in call_func /home/travis/build/neovim/neovim/src/nvim/eval.c:6989:11
    #23 0x7be874 in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:6855:11
    #24 0x7b8cc6 in ex_call /home/travis/build/neovim/neovim/src/nvim/eval.c:2804:9
    #25 0xac0841 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2197:5
    #26 0xa9f407 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #27 0x805730 in ex_execute /home/travis/build/neovim/neovim/src/nvim/eval.c:19299:7
    #28 0xac0841 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2197:5
    #29 0xa9f407 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
SUMMARY: AddressSanitizer: 3 byte(s) leaked in 3 allocation(s).
Runtime errors detected.
```

7.4.2163:

Problem:    match() and related functions tested with old style test.
Solution:   Convert to new style test. (Hirohito Higashi)

https://github.com/vim/vim/commit/d76a0c15f8bdbc901015879177fd5076d34c7a06

7.4.2217:

Problem:    When using matchaddpos() a character after the end of the line can
            be highlighted.
Solution:   Only highlight existing characters. (Hirohito Higashi)

https://github.com/vim/vim/commit/4f416e41243ca151b95d39d81ce23d00b1484755

patch 7.4.2269

Problem:    Using 'hlsearch' highlighting instead of matchpos if there is no
            search match.
Solution:   Pass NULL as last item to next_search_hl() when searching for
            'hlsearch' match. (Shane Harper)

https://github.com/vim/vim/commit/e17bdffff78ebd6a4e3cff26754cc667557ea810

8.0.0033:

Problem:    Cannot use overlapping positions with matchaddpos().
Solution:   Check end of match. (Ozaki Kiichi) Add a test (Hirohito Higashi)

https://github.com/vim/vim/commit/a6c27ee6db2c328e0ab0e6d143e2a295a0bb9c9a

8.0.0040:

Problem:    Whole line highlighting with matchaddpos() does not work.
Solution:   Check for zero length. (Hirohito Higashi)

https://github.com/vim/vim/commit/8507747600bddfd6a68aed057840856bf5548e61

brcolow committed:
vim-patch:7.4.1685

Problem:    There is no easy way to get all the information about a match.
Solution:   Add matchstrpos(). (Ozaki Kiichi)

https://github.com/vim/vim/commit/7fed5c18f8577b75404b80d8b9a9907b1bbd27e4

jamessan commented:
> Okay they are necessary but I am not sure where to put them because list_append_string does not return FAIL if the allocation was unsuccessful.

They aren't necessary.  If any of them fail, nvim exits.  There's another reason it's leaking.  Looking into it.

jamessan commented:
Hmm, `matchlist` does basically the same thing but there are no leaks when it's called... I'm not sure what's going on.

brcolow committed:
vim-patch:7.4.2163

Problem:    match() and related functions tested with old style test.
Solution:   Convert to new style test. (Hirohito Higashi)

https://github.com/vim/vim/commit/d76a0c15f8bdbc901015879177fd5076d34c7a06

brcolow committed:
vim-patch:7.4.2217

Problem:    When using matchaddpos() a character after the end of the line can
            be highlighted.
Solution:   Only highlight existing characters. (Hirohito Higashi)

https://github.com/vim/vim/commit/4f416e41243ca151b95d39d81ce23d00b1484755

brcolow committed:
vim-patch:7.4.2269

Problem:    Using 'hlsearch' highlighting instead of matchpos if there is no
            search match.
Solution:   Pass NULL as last item to next_search_hl() when searching for
            'hlsearch' match. (Shane Harper, closes vim/vim#1013)

https://github.com/vim/vim/commit/e17bdffff78ebd6a4e3cff26754cc667557ea810

brcolow committed:
vim-patch:8.0.0033

Problem:    Cannot use overlapping positions with matchaddpos().
Solution:   Check end of match. (Ozaki Kiichi) Add a test (Hirohito Higashi)

https://github.com/vim/vim/commit/a6c27ee6db2c328e0ab0e6d143e2a295a0bb9c9a

jamessan commented on src/nvim/screen.c:
```diff
@@ -5711,20 +5711,22 @@ next_search_hl_pos(
   shl->lnum = 0;
   for (i = posmatch->cur; i < MAXPOSMATCH; i++) {
-    if (posmatch->pos[i].lnum == 0) {
+    ll_pos_T *pos = &posmatch->pos[i];
```
`s/ll_pos_T/llpos_T/`  I'm telling ya, copy/paste from vim-patch.sh prepared patches is where it's at.  ;)

jamessan commented:
Looks good other than the linting.

brcolow committed:
vim-patch:8.0.0040

Problem:    Whole line highlighting with matchaddpos() does not work.
Solution:   Check for zero length. (Hirohito Higashi)

https://github.com/vim/vim/commit/8507747600bddfd6a68aed057840856bf5548e61
